proposals/0000-support-pun-free-code.md

```diff
@@ -0,0 +1,510 @@
+
+f :: T
+f = D.Proxy
+```
+
+### Module aliases
+
+```haskell
+module M as MyModule, data as D, type as T where
+
+data T = T
+data Mb a = Nth | Jst a
+
+f :: T.T -- will refer to T the type
+f = D.T -- will refer to T the data constructor
+
+g :: Mb T.T -- you don't have to use the aliases in unambiguous situations
+g = D.Jst MyModule.f -- but you can if you do want to
+```
+
+### Punnning Warnings
+
+Recall that `-Wpun-bindings` is triggered at definition sites that use punning,
+and `-Wpuns` is triggered at use sites. To see what qualifies as punning, we
+will look at the code that works today and analyze the breakage that would
+occur if Haskell had a single unified namespace.
+
+#### `-Wpuns`, example #1
+
+```haskell
+module A where { data A = T }
+module B where { data T = X }
+
+module C where
+
+import A
+import B
+
+f = T
+```
+
+If Haskell had a single unified namespace, referring to `T` would result in
+ambiguity (is it `A.T` or `B.T`?), thus this should trigger the warning.
+
+#### `-Wpuns`, example #2
+
+```haskell
+a = 15
+f :: a -> a
+```
+
+If Haskell had a single unified namespace, `a` instead of referring to
+implicitly bound type-variable `a` would refer to `a` on the type-level. This
+means that punning is used and should trigger the warning.
+
+On the contrary:
+
+```haskell
+a = 15
+
+f :: forall a. a -> a
+```
+
+Does not use punning because if Haskell had a single unified namespace, explicitly bound type variable `a` would shadow the top-level `a`.
+
+#### `-Wpuns`, example #3
+
+```haskell
+{-# LANGUAGE ScopedTypeVariabels #-}
+a = 15
+
+f :: forall a. a -> a
+f = \a -> (a :: a)
+--              ^ warning here
+```
+
+In all of the `a` uses except for the last one there is no punning, because if
+Haskell had a single unified namespace, in the type signature, top-level `a`
+would be shadowed by explicitly bound type variable `a`, and in the expression
+`a` variable bound in the lambda would shadow the type variable. In the very
+last case, however, currently, the `a` would refer to the type variable, but if
+Haskell had a single namespace it would refer to the term-level variable. Thus the
+warning is triggered.
+
+#### `-Wpuns`, example #4
+
+```haskell
+f :: [] a -- warning
+g :: [a]  -- warning
+g = []    -- warning
+h = [a]   -- warning
+x = [a,b] -- no warning
+```
+
+Since `Data.BuiltInTypes` is imported by default, all of the cases except the
+very last one will emit `-Wpuns` warning because in all of them it is not clear
+whether the data constructor or a type constructor is being referred to, except
+in the very last case.
+
+#### `-Wpuns`, example #5
+
+```haskell
+f :: ()      -- warning
+f = ()       -- warning
+g :: (a,b)   -- warning
+g = (c,d)    -- warning
+h :: (,) a b -- warning
+h = (,) c d  -- warning
+```
+
+Tuples in this case are very much the same as lists except they will emit a
+warning in all cases.
+
+Note that for both lists and tuples if `-XNoImplicitBuiltInTypes` is enabled,
```

Comment:
I think this is now `-XNoListTupleTypeSyntax` or something similar. Double-check #281.